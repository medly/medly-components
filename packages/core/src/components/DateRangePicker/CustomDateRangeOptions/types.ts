import { Placement } from '../../Popover/types';
import { OptionsProps } from '../../SingleSelect/Options/types';

export interface Props extends OptionsProps {
    placement: Placement;
}
