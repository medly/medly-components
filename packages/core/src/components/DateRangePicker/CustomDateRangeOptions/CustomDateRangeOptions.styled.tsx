import { styled } from '@medly-components/utils';
import { getPosition } from '../../Popover/Popup/styled/Popup.styled';
import Options from '../../SingleSelect/Options';

export const SelectOptions = styled(Options)`
    height: max-content;
    ${getPosition};
    top: ${({ size, placement }) =>
        ((placement === 'bottom' || placement === 'bottom-start' || placement === 'bottom-end') && (size === 'S' ? '4rem' : '5.6rem')) ||
        'unset'};
`;
