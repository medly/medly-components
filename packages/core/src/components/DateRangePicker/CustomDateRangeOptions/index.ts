export { CustomDateRangeOptions as default } from './CustomDateRangeOptions';
