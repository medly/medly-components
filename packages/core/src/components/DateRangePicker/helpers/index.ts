export * from './dateRangeHelpers';
export * from './getFormattedDate';
