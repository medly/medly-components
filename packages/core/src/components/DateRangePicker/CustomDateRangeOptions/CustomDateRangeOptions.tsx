import React from 'react';
import { SelectOptions } from './CustomDateRangeOptions.styled';
import { Props } from './types';

export const CustomDateRangeOptions: React.FC<Props> = React.memo(props => {
    return <SelectOptions {...props} />;
});
CustomDateRangeOptions.displayName = 'CustomDateRangeOptions';
