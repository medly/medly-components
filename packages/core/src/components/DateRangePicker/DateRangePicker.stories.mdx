import { DateRangePicker } from './DateRangePicker';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { boolean, select, text } from '@storybook/addon-knobs';
import {
    placement,
    displayFormats,
    ThemeInterface,
    FormWithDateRangePicker,
    DateRangePickerWithStateForDoc
} from './DateRangePicker.stories';
import { useState } from 'react';
import { DateRangeType, DateRangeSelectionEnum } from './types';

<Meta
    title="Core/DateRangePicker"
    component={DateRangePicker}
    parameters={{
        jest: ['DateRangePicker.test.tsx'],
        design: {
            type: 'figma',
            url: 'https://www.figma.com/file/vA4EQdxl0d0ETtYpw2uSjW/Medly-Pattern-Library-B2B-Internal?node-id=10346%3A1269'
        }
    }}
/>

# DateRangePicker

The DateRangePicker component allows you to select a range of dates through input text or calendar date selection. 

It is a controlled component and must have the following handle state:

```tsx
const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null });
return <DateRangePicker value={dates} onChange={setDates} />;
```

<Preview withToolbar>
    <Story name="Default">
        {() => {
            const [dates, setDates] = useState({ startDate: null, endDate: null });
            return (
                <DateRangePicker
                    id="contract"
                    displayFormat={select('Display Format', displayFormats, 'MM/dd/yyyy')}
                    required={boolean('Required', false)}
                    startDateLabel="From"
                    endDateLabel="To"
                    value={dates}
                    onChange={setDates}
                    size={select('Size', ['S', 'M'], 'M')}
                    popoverPlacement={select('Placement', placement, 'bottom-start')}
                    variant={select('Variant', ['filled', 'outlined'], 'filled')}
                    fullWidth={boolean('Full Width', false)}
                    disabled={boolean('Disabled', false)}
                    withSingleMonth={boolean('With Single Month', false)}
                    errorText={text('Error Text', '')}
                    helperText={text('Helper Text', '')}
                    minSelectableDate={new Date(1901, 0, 1)}
                    maxSelectableDate={new Date(2100, 11, 1)}
                />
            );
        }}
    </Story>
</Preview>

### onPopupClose

To perform any action after the calendar popup closes, pass a function using the `onPopupClose` prop.  You require the `onChange` prop to make dateRangePicker a controlled component and the `onPopupClose` prop to handle actions like API calls that you want to perform after the date selection.
```tsx
import { useAxios } from '@medly-components/utils';
import { createUrlWithParams } from '@utils';

const DateRangePickerWithOnClose: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null }),
        { request } = useAxios();

    const handleApiCall = () =>
        dates.startDate &&
        dates.endDate &&
        request({
            url: createUrlWithParams(url, dates)
            method: 'get'
        });

    return <DateRangePicker value={dates} onChange={setDates} onPopupClose={handleApiCall} />;
};
```

### Validations

1. Validations would happen on invalid and blur event.

```tsx
const FormWithDateRangePicker: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null }),
        handleSubmit = (e: React.FormEvent) => e.preventDefault();
    return (
        <form onSubmit={handleSubmit}>
            <div>
                <DateRangePicker value={dates} onChange={setDates} required />
            </div>
            <Button type="submit">Submit</Button>
        </form>
    );
};
```

<Preview withToolbar>
    <FormWithDateRangePicker />
</Preview>

2. Validations happen only on blur events if you focus out of the component without selecting the date and pass the required prop.


```tsx
const FormWithDatePicker: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null });
    return <DateRangePicker value={dates} onChange={setDates} required />;
};
```

<Preview withToolbar>
    <DateRangePickerWithStateForDoc />
</Preview>

### Custom date range options

Use `customDateRangeOptions` to provide custom date range selection options. For example: 

```tsx
import { DateRangePicker, DateRangeSelectionEnum } from '@medly-components/core';
const DateRangePickerWithCustomDateRangeOptions: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null });
    return <DateRangePicker value={dates} onChange={setDates} customDateRangeOptions={customDateRangeOptions} />;
};
```

<Preview withToolbar>
    <Story name="CustonDateRangeOptions">
        {() => {
            const [dates, setDates] = useState({ startDate: null, endDate: null });
            return (
                <DateRangePicker
                    value={dates}
                    onChange={setDates}
                    displayFormat={select('Display Format', displayFormats, 'MM/dd/yyyy')}
                    required={boolean('Required', false)}
                    size={select('Size', ['S', 'M'], 'M')}
                    popoverPlacement={select('Placement', placement, 'bottom-start')}
                    variant={select('Variant', ['filled', 'outlined'], 'filled')}
                    fullWidth={boolean('Full Width', false)}
                    disabled={boolean('Disabled', false)}
                    withSingleMonth={boolean('With Single Month', false)}
                    errorText={text('Error Text', '')}
                    helperText={text('Helper Text', '')}
                    minSelectableDate={new Date(1901, 0, 1)}
                    maxSelectableDate={new Date(2100, 11, 1)}
                    customDateRangeOptions={[
                        { label: 'Current Week', value: DateRangeSelectionEnum.WEEK },
                        { label: 'Current Month', value: DateRangeSelectionEnum.MONTH },
                        { label: 'Current Quarter', value: DateRangeSelectionEnum.QUARTER },
                        { label: 'Current Year', value: DateRangeSelectionEnum.YEAR },
                        { label: 'Custom', value: DateRangeSelectionEnum.CUSTOM }
                    ]}
                />
            );
        }}
    </Story>
</Preview>

### DateRangePicker Props

<Props of={DateRangePicker} />

### DateRangePicker theme

<Props of={ThemeInterface} />
