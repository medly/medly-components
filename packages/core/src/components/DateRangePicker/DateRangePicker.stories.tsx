import { DateRangePickerTheme, defaultTheme } from '@medly-components/theme';
import React, { useState } from 'react';
import Button from '../Button';
import { Placement } from '../Popover/types';
import { DateRangePicker } from './DateRangePicker';
import { DateRangeProps, DateRangeSelectionEnum, DateRangeType } from './types';

export const placement: Placement[] = [
    'top-start',
    'top',
    'top-end',
    'right-start',
    'right',
    'right-end',
    'bottom-start',
    'bottom',
    'bottom-end',
    'left-end',
    'left',
    'left-start'
];

export const displayFormats: DateRangeProps['displayFormat'][] = [
    'dd/MM/yyyy',
    'MM/dd/yyyy',
    'dd/yyyy/MM',
    'MM/yyyy/dd',
    'yyyy/dd/MM',
    'yyyy/MM/dd',
    'dd-MM-yyyy',
    'dd-yyyy-MM',
    'MM-dd-yyyy',
    'MM-yyyy-dd',
    'yyyy-dd-MM',
    'yyyy-MM-dd'
];

export const CustomDateRangeOptions = [
    { label: 'Current Week', value: DateRangeSelectionEnum.WEEK },
    { label: 'Current Month', value: DateRangeSelectionEnum.MONTH },
    { label: 'Current Quarter', value: DateRangeSelectionEnum.QUARTER },
    { label: 'Current Year', value: DateRangeSelectionEnum.YEAR },
    { label: 'Custom', value: DateRangeSelectionEnum.CUSTOM }
];

export const ThemeInterface: React.FC<DateRangePickerTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.dateRangePicker
};

export const FormWithDateRangePicker: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null }),
        handleSubmit = (e: React.FormEvent) => e.preventDefault();
    return (
        <form onSubmit={handleSubmit}>
            <div>
                <DateRangePicker value={dates} onChange={setDates} required />
            </div>
            <Button type="submit">Submit</Button>
        </form>
    );
};

export const DateRangePickerWithStateForDoc: React.FC = () => {
    const [dates, setDates] = useState<DateRangeType>({ startDate: null, endDate: null });
    return <DateRangePicker value={dates} onChange={setDates} required />;
};
