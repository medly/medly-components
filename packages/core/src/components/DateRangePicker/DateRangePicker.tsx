import { useOuterClickNotifier, useUpdateEffect } from '@medly-components/utils';
import React, { FC, useCallback, useRef, useState } from 'react';
import * as TextFieldStyled from '../TextField/Styled';
import { CustomDateRangeOptions } from './CustomDateRangeOptions/CustomDateRangeOptions';
import DateRangeCalendar from './DateRangeCalendar';
import DateRangeTextFields from './DateRangeTextFields';
import { dateRangeHelpers } from './helpers/dateRangeHelpers';
import { DateRangeProps, DateRangeSelectionEnum, PopoverTypes } from './types';

export const DateRangePicker: FC<DateRangeProps> = React.memo(props => {
    const {
        id,
        value,
        minWidth,
        fullWidth,
        displayFormat,
        startDateLabel,
        endDateLabel,
        errorText,
        helperText,
        variant,
        disabled,
        showDecorators,
        size,
        onBlur,
        minSelectableDate,
        maxSelectableDate,
        popoverPlacement,
        onChange,
        required,
        validator,
        withSingleMonth,
        showTooltipForHelperAndErrorText,
        customDateRangeOptions,
        ...restProps
    } = props;
    const startDateRef = useRef<HTMLInputElement>(null),
        endDateRef = useRef<HTMLInputElement>(null),
        wrapperRef = useRef<HTMLDivElement>(null),
        [isActive, setActive] = useState(false),
        [activePopover, setActivePopover] = useState<PopoverTypes>(PopoverTypes.CALENDAR),
        wrapperMinWidth = minWidth ? minWidth : customDateRangeOptions.length ? '38rem' : '33.8rem',
        [focusedElement, setFocusedElement] = useState<'START_DATE' | `END_DATE`>('START_DATE'),
        focusElement = useCallback(element => (element === 'START_DATE' ? startDateRef : endDateRef).current.focus(), []),
        onCustomRangeIconClick = useCallback(() => {
            if (!disabled) {
                activePopover === PopoverTypes.CUSTOM_RANGE_OPTIONS ? setActive(value => !value) : setActive(true);
                setActivePopover(PopoverTypes.CUSTOM_RANGE_OPTIONS);
            }
        }, [disabled, activePopover]),
        onCalendarIconClick = useCallback((value: boolean) => {
            setActivePopover(PopoverTypes.CALENDAR);
            setActive(value);
        }, []),
        onOptionClick = useCallback(
            (option: any) => {
                if (option.value === DateRangeSelectionEnum.CUSTOM) {
                    focusElement('START_DATE');
                    setActive(true);
                } else {
                    const selectedRange = dateRangeHelpers[
                        option.value as Exclude<DateRangeSelectionEnum, DateRangeSelectionEnum.CUSTOM>
                    ]();
                    if (minSelectableDate && selectedRange.startDate < minSelectableDate) {
                        selectedRange.startDate = minSelectableDate;
                    }
                    if (maxSelectableDate && selectedRange.endDate > maxSelectableDate) {
                        selectedRange.endDate = maxSelectableDate;
                    }
                    onChange && onChange(selectedRange);
                    setActive(false);
                }
            },
            [onChange, focusElement]
        );

    useOuterClickNotifier(() => {
        setActive(false);
    }, wrapperRef);

    useUpdateEffect(() => focusElement(focusedElement), [focusedElement]);

    return (
        <TextFieldStyled.OuterWrapper id={`${id}-wrapper`} ref={wrapperRef} fullWidth={fullWidth} minWidth={wrapperMinWidth} {...restProps}>
            <DateRangeTextFields
                id={id}
                size={size}
                required={required}
                variant={variant}
                errorText={errorText}
                helperText={helperText}
                disabled={disabled}
                showDecorators={showDecorators}
                isActive={isActive}
                activePopover={activePopover}
                validator={validator}
                startDateLabel={startDateLabel}
                endDateLabel={endDateLabel}
                selectedDates={value}
                onDateChange={onChange}
                displayFormat={displayFormat}
                onCalendarIconClick={onCalendarIconClick}
                setFocusedElement={setFocusedElement}
                startDateRef={startDateRef}
                endDateRef={endDateRef}
                onBlur={onBlur}
                showTooltipForHelperAndErrorText={showTooltipForHelperAndErrorText}
                onCustomRangeIconClick={onCustomRangeIconClick}
                showOptions={!!customDateRangeOptions.length}
            />
            {isActive &&
                (activePopover === PopoverTypes.CALENDAR ? (
                    <DateRangeCalendar
                        id={`${id}-calendar`}
                        size={size}
                        placement={popoverPlacement}
                        selectedDates={value}
                        setActive={onCalendarIconClick}
                        withSingleMonth={withSingleMonth}
                        focusElement={focusElement}
                        onDateSelection={onChange}
                        focusedElement={focusedElement}
                        setFocusedElement={setFocusedElement}
                        minSelectableDate={minSelectableDate}
                        maxSelectableDate={maxSelectableDate}
                    />
                ) : (
                    <CustomDateRangeOptions
                        id={`${id}-custom-date-range-options`}
                        size={size}
                        variant={variant}
                        onOptionClick={onOptionClick}
                        options={customDateRangeOptions}
                        placement={popoverPlacement}
                    />
                ))}
        </TextFieldStyled.OuterWrapper>
    );
});

DateRangePicker.displayName = 'DateRangePicker';
DateRangePicker.defaultProps = {
    id: 'medly-date-range-picker',
    size: 'M',
    variant: 'filled',
    startDateLabel: 'From',
    endDateLabel: 'To',
    displayFormat: 'MM-dd-yyyy',
    popoverPlacement: 'bottom-start',
    withSingleMonth: false,
    minSelectableDate: new Date(1901, 0, 1),
    maxSelectableDate: new Date(2100, 11, 1),
    showDecorators: true,
    customDateRangeOptions: []
};
