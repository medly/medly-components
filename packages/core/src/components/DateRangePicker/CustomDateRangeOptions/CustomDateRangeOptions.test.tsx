import { render } from '@test-utils';
import React from 'react';
import CustomDateRangeOptions from '.';
import { CustomDateRangeOptions as SelectOptions } from '../DateRangePicker.stories';

describe('CustomDateRangeOptions', () => {
    it('should render properly ', () => {
        const { container } = render(
            <CustomDateRangeOptions
                id="contract-custom-date-range-options"
                size="S"
                variant="outlined"
                options={SelectOptions}
                placement="top"
                onOptionClick={jest.fn()}
            />
        );

        expect(container.querySelector('#contract-custom-date-range-options')).toMatchSnapshot();
    });
});
