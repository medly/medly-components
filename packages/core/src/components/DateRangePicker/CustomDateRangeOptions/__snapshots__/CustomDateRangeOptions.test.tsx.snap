// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDateRangeOptions should render properly  1`] = `
.c4 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c2 {
  cursor: pointer;
  padding: 0 0.8rem;
  border-radius: 0.4rem;
  box-sizing: border-box;
  min-height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  background-color: #ffffff;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
}

.c2:hover {
  background-color: #F7F8F9;
}

.c2:active {
  background-color: #eff2f4;
}

.c2 > .c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  color: #13181D;
}

.c2 > .sc-AxirZ {
  margin-left: 1.6rem;
}

.c2 > .sc-AxirZ * {
  fill: #13181D;
}

.c0 {
  z-index: 4;
  position: absolute;
  background: #ffffff;
  top: 4rem;
  left: 0;
  margin: 0;
  padding: 0.8rem;
  width: 100%;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  list-style: none;
  box-sizing: border-box;
  box-shadow: 0 0.2rem 0.8rem #B0BCC8;
  border-radius: 0.4rem;
  overflow: auto;
  max-height: 20rem;
}

.c1 {
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  bottom: calc(100% + 0px);
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  top: unset;
}

<ul
  class="c0 c1"
  id="contract-custom-date-range-options"
>
  <li
    class="c2"
    id="Current-Week"
  >
    <span
      class="c3 c4"
    >
      Current Week
    </span>
  </li>
  <li
    class="c2"
    id="Current-Month"
  >
    <span
      class="c3 c4"
    >
      Current Month
    </span>
  </li>
  <li
    class="c2"
    id="Current-Quarter"
  >
    <span
      class="c3 c4"
    >
      Current Quarter
    </span>
  </li>
  <li
    class="c2"
    id="Current-Year"
  >
    <span
      class="c3 c4"
    >
      Current Year
    </span>
  </li>
  <li
    class="c2"
    id="Custom"
  >
    <span
      class="c3 c4"
    >
      Custom
    </span>
  </li>
</ul>
`;
